src/python/gudhi/simplex_tree.pyx

```diff
@@ -797,6 +799,37 @@ cdef class SimplexTree:
         :rtype: bool
         """
         return dereference(self.get_ptr()) == dereference(other.get_ptr())
+    
+    def __getstate__(self):
+        """Pickle the SimplexTree data structure as a Python Byte Array
+        :raises MemoryError: In the case the serialization allocates a too large block of memory.
+        :returns: Serialized SimplexTree data structure
+        :rtype: Byte Array
+        """
+        cdef char* buffer
+        cdef size_t buffer_size = 0
+        with nogil:
+            buffer = self.get_ptr().serialize(buffer_size)
+        
+        return PyByteArray_FromStringAndSize(buffer, buffer_size)
```

Comment:
Now that we have a function that can compute the size before doing the serialization, it is really tempting to do the allocation through Python (`np.empty(thesize, dtype='B')` for instance) and pass that buffer to C++. This makes fewer copies, and avoids the complicated call to `delete[]` in cython.

Reply:
## With a buffer size of 2^33
### c++
```cpp
char* buff = new char[[size](std::pow(2, 33))]; delete[] buff; // in: 0s
std::string str; str.resize([size](std::pow(2, 33))); // in: 1.96s - that is pretty bad...
std::stringview sv; // std::stringview has no resize method ?!...
```
### python
```python
import numpy as np
from timeit import default_timer as timer
from datetime import timedelta
start = timer(); a = np.empty(pow(2,33), dtype='B'); del a; print(timedelta(seconds=timer()-start)) # in: 0.000098 sec.
```

Reply:
This involves quite a lot of changes that I did on [82b3c2a](https://github.com/GUDHI/gudhi-devel/pull/810/commits/82b3c2aa529912547ff16dbc09348979010ba477). I first tried with `std::string`, but benchmark was not ok. So I rolled back and use the numpy method you are mentionning and benchmarked it. Almost the same results, but with no more memory leaks.

Reply:
Here, `new char[1LL<<33]` and `np.empty(2**33)` indeed take essentially no time (linux delays actually allocating the pages until they are accessed), while `new char[1LL<<33]()`, `str.resize(1LL<<33)` and `np.zeros(2**33)` take 0.7s since they actually write to the whole buffer (a later call to memset takes only .25s, so a large part of the .7s is the delayed allocation).
(to confuse things, `calloc` also takes almost no time, but let's ignore that)
`string_view` is a view into memory it does not own, it seems normal not to have a resize method that could only shrink.